sqlgen/matcher.go

```diff
@@ -0,0 +1,138 @@
+//
+// Naively, to match a row against these queries, you would have to test all
+// queries, which is inefficient for large sets of queries.
+//
+// Instead, the matcher uses the fact that the queries id=10 and id=20 have the
+// same shape to optimize matching them: The matcher extracts the common set of
+// columns, (id), and the specific values matched by each query, (10) and (20).
+// Then, to match a query, the matcher extracts the id from that query, and
+// looks it up in a lookup table from tuples to query. Here, that lookup table
+// will hold (10) and (20) associated with some opaque id (such as a *query
+// pointer, or some index in array) to identify the queries.
+//
+// For every different query shape, the matcher builds a *matcherGroup holding
+// a lookup table, letting the matcher process queries in time O(#query shapes).
+type matcher struct {
+	groups map[string]*matcherGroup
+}
+
+// A matcherGroup holds a set of tuples for a given set of columns in a
+// matcher.
+type matcherGroup struct {
+	columns []string
+	// tuples is a map from column values to sets of queries interested in those
+	// specific values. For example, if columns is []string{"city", "name"},
+	// tuples is a map from tuples such as [2]interface{}{"sf", "bob"} and
+	// [2]interface{}{"sf", "alice"} to sets of opaque query handles.
+	tuples map[interface{}]map[interface{}]struct{}
+}
+
+// newMatcher creates a new matcher.
+func newMatcher() *matcher {
+	return &matcher{
+		groups: make(map[string]*matcherGroup),
+	}
+}
+
+// add adds a query to the matcher, associating an opaque id with the query.
+func (m *matcher) add(id interface{}, filter Filter) {
+	columnSet := extractColumns(filter)
+	values := internal.ToArray(extractColumnValues(filter, columnSet))
+	key := columnsKey(columnSet)
+
+	g, ok := m.groups[key]
+	if !ok {
+		g = &matcherGroup{
+			columns: columnSet,
+			tuples:  make(map[interface{}]map[interface{}]struct{}),
+		}
+		m.groups[key] = g
+	}
+
+	tuples, ok := g.tuples[values]
+	if !ok {
+		tuples = make(map[interface{}]struct{})
+		g.tuples[values] = tuples
+	}
+
+	tuples[id] = struct{}{}
+}
+
+// removes removes a query from the matcher.
+func (m *matcher) remove(id interface{}, filter Filter) {
+	columnSet := extractColumns(filter)
+	values := internal.ToArray(extractColumnValues(filter, columnSet))
+	key := columnsKey(columnSet)
+
+	g, ok := m.groups[key]
+	if !ok {
+		return
+	}
+
+	tuples, ok := g.tuples[values]
+	if !ok {
+		return
+	}
+
+	delete(tuples, id)
+	if len(tuples) > 0 {
+		return
+	}
+
+	delete(g.tuples, values)
+	if len(g.tuples) > 0 {
+		return
+	}
+
+	delete(m.groups, key)
+}
+
+// match matches an object against all queries and returns all the opaque
+// values associated with matching queries.
+func (m *matcher) match(o map[string]interface{}) []interface{} {
+	var result []interface{}
+	for _, g := range m.groups {
+		for value := range g.tuples[internal.ToArray(extractColumnValues(o, g.columns))] {
```

Comment:
consider `s/value/id/` (and maybe updating the comment) to match the `add` function.